macOS/DuckDuckGo/DataImport/Bookmarks/FavoritesUtils.swift

```diff
@@ -0,0 +1,96 @@
+//
+//  FavoritesUtils.swift
+//
+//  Copyright © 2025 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+final class FavoritesUtils {
+
+    /// Merges imported bookmarks with the provided array of favorites, marking bookmarks with matching URLs as favorites.
+    /// Any favorites not matching existing bookmarks will be added to the bookmark bar.
+    static func mergeBookmarksAndFavorites(bookmarks: ImportedBookmarks, favorites: [ImportedBookmarks.BookmarkOrFolder]) -> ImportedBookmarks {
+        guard !favorites.isEmpty else {
+            return bookmarks
+        }
+
+        // Recursively process bookmarks to mark matching URLs as favorites
+        func processBookmarkOrFolder(_ item: ImportedBookmarks.BookmarkOrFolder) -> ImportedBookmarks.BookmarkOrFolder {
+            var updatedItem = item
+
+            if !item.isFolder, let shortcut = favorites.first(where: { $0.urlString == item.urlString }) {
```

Comment:
should we instead make a `Set(favorites.map(\.urlString))` for faster further checks?